### DISCUSSION TITLE
The coordinates are in millimeters in PyMuPDF

### DISCUSSION THREAD

**Discussion by Sv9t** [pymupdf/PyMuPDF#797] [Q&A] [answered]
Hello!
I have coordinates of an image  in millimeters in pdf file. Coordinates of the top-left corner: x=24mm, y=88mm. And I have the width of the image: 77mm, height image: 7mm.
How I can use my coordinates in millimeters in function fitz.Rect()? Rect() uses points in coordinates on the pdf page. I was trying to convert millimeters to pixels (point), but this method did not help.

**Comment by JorjMcKie** [comment]
You should be able to compute mm by using these equalities: 72 points = 1 inch, 1 inch = 25.4 mm.
What did you use, that did not work?
Your point x=24mm, y=88mm should therefore compute to x: 24 / 25.4 * 72 = 68 points, y: 88 / 25.4 * 72 = 304,9 point, etc.
Counter-calculation: page size ISO A4 = 595 x 842 points => 595/72*2.54 = 21 cm width, and 842/72*2.54 = 29.7 cm.

**Reply by JorjMcKie** [reply]
oops, GitHub editor tricked me out: ``595 / 72 * 2.54`` and ``842 / 72 * 2.54``!

**Comment by Sv9t** [comment]
> What did you use, that did not work?

my bad, I convert millimeters in pixels, from the calculation 1mm=3.793627px. This is wrong!

> 595/722.54 = 21 cm width, and 842/722.54 = 29.7 cm.

I didn't understand where the number came from 722.54?
![pymupdf](https://user-images.githubusercontent.com/14145634/103302757-70abad80-4a15-11eb-9668-adf778d9943e.jpg)


I drew my calculations, attach jpeg. If use fitz.Rect(68, 304.9, 308.4, 517,3) then I get an error "raise ValueError("bad Rect: sequ. length") ValueError: bad Rect: sequ. length" 
Please tell me what I calculated wrong?

**Reply by JorjMcKie** [reply]
rectangle constructor needs 4 numbers - you gave 5!

**Reply by JorjMcKie** [reply]
probably _**517.3**_ - you coded a comma.

**Reply by Sv9t** [reply]
Thanks! This is my mistake! I wasn't attentive
The image insert to pdf page, but the image is much lower for X coordinate than it should be. I'll try different options...

**Reply by JorjMcKie** [reply]
>  The image insert to pdf page, but the image is much lower for X coordinate than it should be. I'll try different options...

There may be [this](https://pymupdf.readthedocs.io/en/latest/faq.html#misplaced-item-insertions-on-pdf-pages) issue, so that you must "wrap" the page contents first.

**Reply by JorjMcKie** [reply]
just try `page.wrap_contents()` before you insert.

**Comment by Sv9t** [comment]
```python
import fitz

file = '56204820.pdf'
regnum = 'regnum.png'
# Coordinates of the top-left corner: x=24mm, y=88mm
# Width of the image: 77mm and height image: 7mm.
coords = [24, 88, 77, 7]

def func_coords(*args):
    print(args)
    point_coords = []
    for c in args[0][:2]:
        point_coords.append(c / 25.4 * 72)

    for c in args[0][2:]:
        points = c / 25.4 * 72
        if c == args[0][2]:
            result = 595 - points - args[0][0]
        else:
            result = 842 - points - args[0][1]
        point_coords.append(result)
    return tuple(point_coords)


def main():
    pdf = fitz.open(file)
    page = pdf[0]

    point_coords = func_coords(coords)
    print(point_coords)

    rect = fitz.Rect(point_coords)
    img = open(regnum, "rb").read()  # an image file
    page.wrapContents()
    page.insertImage(rect, stream=img)

    pdf.save('result.pdf')


if __name__ == "__main__":
    main()
```

This is my script, maybe there is a problem in the code? String _page.wrapContents()_ it didn't help me, image is much lower for Y coordinate than it should be.

**Comment by JorjMcKie** [comment, marked as answer]
A rectangle must be given as `rect = fitz.Rect(x0, y0, x1, y1)` where (x0, y0) is the top-left point and (x1, y1) the bottom-right one. If you have the the top-left only plus the width / height, the calculation is therefore `rect = fitz.Rect(x0, y0, x0 + width, y0 + height)`.
So in your case `rect = fitz.Rect(24, 88, 24+77, 88+7)`. Then simply do this to convert the whole thing to pixels:
```python
>>> rect = fitz.Rect(24, 88, 24+77, 88+7)
>>> rect = rect / 25.4 * 72
>>> rect
Rect(68.03149606299213, 249.4488188976378, 286.2992125984252, 269.29133858267716)
>>> 
```

**Comment by Sv9t** [comment]
Excellent!! It works! 
I calculated rectangle  the bottom-right wrong, I calculated from the bottom-right corner of the a4 page, but I had to calculate from the top-left corner.
Thank you for your help and your feedback! PyMuPDF it is very necessary module!
